openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -130,45 +247,166 @@ def __contains__(self, item):
+
+
 
-    _VALENCE_TYPE = None  # ChemicalEnvironment valence type for checking SMIRKS is conformant
 
     # TODO: Allow preferred units for each parameter type to be specified and remembered as well for when we are writing out
 
     # TODO: Can we provide some shared tools for returning settable/gettable attributes, and checking unit-bearing attributes?
 
-    def __init__(self, smirks=None, **kwargs):
+    def __init__(self, smirks=None, permit_cosmetic_attributes=False, **kwargs):
         """
         Create a ParameterType
 
         Parameters
         ----------
         smirks : str
             The SMIRKS match for the provided parameter type.
+        permit_cosmetic_attributes : bool optional. Default = False
+            Whether to store non-spec kwargs as "cosmetic attributes", which can be accessed and written out.
 
         """
+        import openforcefield.utils.toolkits
+
+        self._COSMETIC_ATTRIBS = []  # A list that may be populated to record the cosmetic
+        # attributes read from a SMIRNOFF data source
+
         if smirks is None:
             raise ValueError("'smirks' must be specified")
+
+        # TODO: Make better switch for toolkit registry
+        if openforcefield.utils.toolkits.OPENEYE_AVAILABLE:
+            toolkit = 'openeye'
+        elif openforcefield.utils.toolkits.RDKIT_AVAILABLE:
+            toolkit = 'rdkit'
+        ChemicalEnvironment.validate(
+            smirks, ensure_valence_type=self._VALENCE_TYPE, toolkit=toolkit)
+
         self._smirks = smirks
 
-        # Handle all unknown kwargs as cosmetic so we can write them back out
+
+        # First look for indexed attribs, removing them from kwargs as they're found
+        for unidx_key in self._INDEXED_ATTRIBS:
+            index = 1
+            idx_key = unidx_key+str(index)
+            if idx_key in kwargs:
+                setattr(self, unidx_key, list())
+            while idx_key in kwargs:
+                val = kwargs[idx_key]
+                if unidx_key in self._REQUIRE_UNITS:
+                    if not val.unit.is_compatible(self._REQUIRE_UNITS[unidx_key]):
+
+                        msg = "{} constructor received kwarg {} with value {}, " \
+                              "which is incompatible with expected unit {}".format(self.__class__,
+                                                                                   idx_key,
+                                                                                   val,
+                                                                                   self._REQUIRE_UNITS[unidx_key])
+                        raise SMIRNOFFSpecError(msg)
+                if unidx_key in self._ATTRIBS_TO_TYPE:
+                    type_to_cast = self._ATTRIBS_TO_TYPE[unidx_key]
+                    val = type_to_cast(val)
+                getattr(self, unidx_key).append(val)
+                del kwargs[idx_key]
```

Comment:
I'm not sure if this matters but you may want to consider the case where someone accidentally misses an index (e.g param1, param3, param4) in the kwargs dict - I think currently the code would just stop when it hit the gap?

Reply:
If it helps, you could use a regex to find all the `param_name+number` strings as suggested here: https://stackoverflow.com/a/51174978 .

Reply:
> I think currently the code would just stop when it hit the gap?

So, what will happen right now if there's a gap (eg. `k1`, `k2`, and `k4` are defined) is that `k1` and `k2` will get set into a list `self.k`, with length 2. `k1` and `k2` will, in the process, be removed from `kwargs`. However, `k4` won't be removed from `kwargs`, and will get sent into the non-indexed kwarg handling code. That code won't recognize it, and will throw an exception. 

I've implemented a test for this scenario here: https://github.com/openforcefield/openforcefield/blob/ff-offxml-io/openforcefield/tests/test_parameters.py#L454-L470

> If it helps, you could use a regex to find all the param_name+number strings

Maybe it's just me, but I've had a lot of trouble with regexes in code. They're not all that readable, and can make debugging really complicated. I'm not totally against using them, but I would prefer to save them for cases where they're absolutely needed.